src/shim/shim.c

```diff
@@ -298,6 +302,125 @@ static void _shim_parent_init_ptrace() {
     }
 }
 
+static void _handle_sigsys(int sig, siginfo_t* info, void* voidUcontext) {
+    ucontext_t* ctx = (ucontext_t*)(voidUcontext);
+    if (sig != SIGSYS) {
+        abort();
+    }
+    // Make the syscall via the *the shim's* syscall function (which overrides
+    // libc's).  It in turn will either emulate it or (if interposition is
+    // disabled), or make the call natively. In the latter case, the syscall
+    // will be permitted to execute by the seccomp filter.
+    long rv = shadow_raw_syscall(ctx->uc_mcontext.gregs[REG_RAX], ctx->uc_mcontext.gregs[REG_RDI],
+                                 ctx->uc_mcontext.gregs[REG_RSI], ctx->uc_mcontext.gregs[REG_RDX],
+                                 ctx->uc_mcontext.gregs[REG_R10], ctx->uc_mcontext.gregs[REG_R8],
+                                 ctx->uc_mcontext.gregs[REG_R9]);
+    ctx->uc_mcontext.gregs[REG_RAX] = rv;
+}
+
+static void _shim_parent_init_seccomp() {
+    // Install signal sigsys signal handler, which will receive syscalls that
+    // get stopped by the seccomp filter.
+    if (sigaction(SIGSYS,
+                  &(struct sigaction){
```

Comment:
Does it make sense to check for an existing `SIGSYS` handler first? It might make sense to emit a warning in that case.

Reply:
> Does it make sense to check for an existing SIGSYS handler first? It might make sense to emit a warning in that case.

Done

> Is there anything nice we can do if an application tries to do its own sandboxing and overwrites the sigsys signal handler (for example if someone runs tor with sandboxing enabled)?

I *think* the virtual process using seccomp itself won't *necessarily* interfere with our usage. Additional filters get added on; old ones never get removed. (I suppose if it blocks the syscalls we use for IPC things would break)

We would have a problem if our SIGSYS handler got blocked or overridden, though. For now I implemented handlers in the related syscalls to prevent that from happening. We *might* be able to handle an attempt to set a sigaction more transparently by having our handler also call the specified handler, but probably not worth it right now.